pyomo/contrib/pynumero/src/hsl_interface/ma27Interface.c

```diff
@@ -0,0 +1,242 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+void abort_bad_memory(int status){
+	printf("Bad memory allocation in MA27 C interface. Aborting.");
+	exit(status);
+}
+
+struct MA27_struct {
+	int LIW_a, LIW_b, NSTEPS, IFLAG, LA, MAXFRT;
+	double IW_factor, A_factor;
+	bool A_allocated, IKEEP_allocated;
+	bool IW_a_allocated, IW_b_allocated;
+	int* IW_a;
+	int* IW_b;
+	// Use different arrays for IW that is sent to MA27A and that sent to
+	// MA27B because IW must be discarded after MA27A but kept after MA27B.
+	// If these arrays are the same, and a symbolic factorization is performed
+	// after a numeric factorization (e.g. on a new matrix), user-defined
+	// and MA27B-defined allocations of IW can be conflated.
+	int* IW1;
+	int* IKEEP;
+	int ICNTL[30], INFO[20];
+	double OPS;
+	double* W;
+	double* A;
+	double CNTL[5];
+};
+
+struct MA27_struct* new_MA27_struct(void){
+
+	struct MA27_struct* ma27 = malloc(sizeof(struct MA27_struct));
+	if (ma27 == NULL) { abort_bad_memory(1); }
+
+	ma27id_(ma27->ICNTL, ma27->CNTL);
+
+	// Set default values of parameters
+	ma27->A_allocated = ma27->IKEEP_allocated = false;
+	ma27->IW_a_allocated = ma27->IW_b_allocated = false;
+	ma27->IFLAG = 0;
+	ma27->IW_factor = 1.2;
+	ma27->A_factor = 2.0;
+
+	// Return pointer to ma27 that Python program can pass to other functions
+	// in this code
+	return ma27;
+}
+
+// Functions for setting/accessing INFO/CNTL arrays:
+void set_icntl(struct MA27_struct* ma27, int i, int val) {
+  ma27->ICNTL[i] = val;
+}
+int get_icntl(struct MA27_struct* ma27, int i) {
+	return ma27->ICNTL[i];
+}
+void set_cntl(struct MA27_struct* ma27, int i, double val) {
+	ma27->CNTL[i] = val;
+}
+double get_cntl(struct MA27_struct* ma27, int i) {
+	return ma27->CNTL[i];
+}
+int get_info(struct MA27_struct* ma27, int i) {
+	return ma27->INFO[i];
+}
+
```

Comment:
Can the number of negative eigenvalues be gotten from get_info?

Reply:
Yes, it is INFO(15) and (24) in MA27 and 57 respectively